apply user defined tags to all resources (#706)
openshift/enhancements: enhancements/api-review/custom-tags-aws.md

   list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other
   bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster.
 - tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
   after some period of time, apply the tags post-creation.

### Non-Goals

 - to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
   changes the tags stored in the infrastructure resource, behavior is undefined. See below.
 - to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
   to prohibit that later.

## Proposal

New `experimentalPropagateUserTags` field added to `.platform.aws` of install config to indicate that the user tags should be applied to AWS
resources created by in-cluster operators.

If `experimentalPropagateUserTags` is true, install validation will fail if there is any tag that starts with `kubernetes.io` or `openshift.io`.

> Add a new field `userTags` to `.status.aws` of the `infrastructure.config.openshift.io` type. Tags included in the

Reviewer: I disagree with putting these values in status.  For 1, it will make it more difficult for users to change if they need to for bug/support purposes.  For 2, if someone attempts to recreate this object from a backup or source control repo, oc apply will not add the elements to status.  Thus, this could create a situation where users have trouble restoring something after accidental deletion.

User defined values belong in spec.

Reviewer 2: Is status not preserved on this field today with velero backup?  We have other write once status fields.

Reviewer 2: etcd backup restore will preserve this as well.

Reviewer 3: Looking at this issue, I don't think velero restores the status no: https://github.com/vmware-tanzu/velero/issues/1687 

I think in general there is an assumption that a resources status should be computable by the controller that reconciles it, though that doesn't apply when your CR is informational like this one is

Reviewer 2: ok, it can be added to spec at a future time.  Present backup and restore is at etcd layer itself.  we can evolve this part post 4.8

`userTags` field will be applied to new resources created for the cluster. The `userTags` field will be populated by the installer only if the `experimentalPropagateUserTags` field is true.

Installer will apply these tags to all AWS resources it creates with terraform (e.g. bootstrap and master EC2 instances) (already exists)

All operators that create AWS resources (ingress, cloud credential, storage, image registry, machine-api)
will apply these tags to all AWS resources they create.

userTags that are specified in the infrastructure resource will merge with userTags specified in an individual component. In the case where a userTag is specified in the infrastructure resource and there is a tag with the same key specified in an individual component, the value from the individual component will have precedence and be used.

### User Stories

https://issues.redhat.com/browse/SDE-1146 - IAM users and roles can only operate on resources with specific tags
As a security-conscious ROSA customer, I want to restrict the permissions granted to Red Hat in my AWS account by using
AWS resource tags.

### Risks and Mitigations

> The userTags field is intended to be set at install time and immutable. If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. If tags are removed from userTags, the tag may or may not be removed from the AWS resource.

Reviewer: I would like to see some of the expected scenarios documented during development/QA for each component that responds to these tags.

---
title: apply-user-defined-tags-to-all-aws-resources-created-by-openshift
authors:
  - "@gregsheremeta"
reviewers:

>   - @decarr @bparees

Author: fixed, thanks

approvers:
  - @decarr 
creation-date: 2021-03-24
last-updated: 2021-04-09
status: implementable
---

# Apply user defined tags to all AWS resources created by OpenShift

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement describes a proposal to allow an administrator of OpenShift to

> have the ability to apply user defined tags to all resources created by OpenShift in AWS.

Reviewer 4: ```suggestion
have the ability to apply user defined tags to many resources created by OpenShift in AWS.
```

Author: done, thanks

 - tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
   after some period of time, apply the tags post-creation.

### Non-Goals

 - to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
   changes the tags stored in the infrastructure resource, behavior is undefined. See below.
 - to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
   to prohibit that later.

## Proposal

New `experimentalPropagateUserTags` field added to `.platform.aws` of install config to indicate that the user tags should be applied to AWS
resources created by in-cluster operators.

If `experimentalPropagateUserTags` is true, install validation will fail if there is any tag that starts with `kubernetes.io` or `openshift.io`.

Add a new field `userTags` to `.status.aws` of the `infrastructure.config.openshift.io` type. Tags included in the
`userTags` field will be applied to new resources created for the cluster. The `userTags` field will be populated by the installer only if the `experimentalPropagateUserTags` field is true.

> Installer will apply these tags to all AWS resources it creates with terraform (e.g. bootstrap and master EC2 instances) (already exists)

Reviewer 4: clarification:  the installer applies the tags from the installconfig, not from infra status, and not based on the propagation option being set.  This is existing behavior.

Author: done, thanks

All operators that create AWS resources (ingress, cloud credential, storage, image registry, machine-api)
will apply these tags to all AWS resources they create.

userTags that are specified in the infrastructure resource will merge with userTags specified in an individual component. In the case where a userTag is specified in the infrastructure resource and there is a tag with the same key specified in an individual component, the value from the individual component will have precedence and be used.

The userTags field is intended to be set at install time and is considered immutable. Components that respect this field must only ever add tags that they retrieve from this field to cloud resources, they must never remove tags from the existing underlying cloud resource even if the tags are removed from this field(despite it being immutable).

> If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. If tags are removed from userTags, the tag may or may not be removed from the AWS resource.

Reviewer 4: ```suggestion
If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. However if tags are removed from userTags, the tag must not be removed from the AWS resource.
```

Author: done, thanks

### User Stories

https://issues.redhat.com/browse/SDE-1146 - IAM users and roles can only operate on resources with specific tags
As a security-conscious ROSA customer, I want to restrict the permissions granted to Red Hat in my AWS account by using
AWS resource tags.

### Risks and Mitigations

## Design Details

### Test Plan

TBD

### Graduation Criteria

TBD

> ### Upgrade / Downgrade Strategy

Reviewer 4: On upgrade:
The new status field won't be populated since it is only populated by the installer and that can't have happened if the cluster was installed from a prior version.  Components that consume the new field should take no action since they will see no additional tags

On downgrade:
The status field may remain populated(i'm not actually sure what will get served from a CRD api perspective), components may or may not continue to tag newly created resources w/ the additional tags depending on whether or not a given component still has logic to respect the status tags, after the downgrade.

Reviewer 4: (the above is a suggested entry for the `upgrade/downgrade strategy` section)

Reviewer 2: Agree with above  , less concerned during experimental.

Author: added, thanks!

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade in order to keep previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade in order to make use of the enhancement?

Good question ^. TBD

### Version Skew Strategy

TBD

## Implementation History

> ## Drawbacks

Reviewer 4: If a customer decides they do not want these tags applied to new resources there is no clean way to address that need:

1) they would have to get help from support to edit the status field
2) they'd have to manually remove the undesired tags from any existing resources

In the future we will want to introduce a spec field where a customer can specify+edit these tags, which will be reflected into status when changed, and we will expect consuming components to reconcile changes to the set of tags by applying net new tags to existing resources (but they will still not remove tags that are dropped from the list of tags)

Reviewer 4: (the above is a suggested entry for the `drawbacks` section)

Author: added, thanks!

---
title: apply-user-defined-tags-to-all-aws-resources-created-by-openshift
authors:
  - "@gregsheremeta"
reviewers:
  - @decarr
  - @bparees
approvers:
  - @decarr 
creation-date: 2021-03-24
last-updated: 2021-04-09
status: implementable
---

# Apply user defined tags to all AWS resources created by OpenShift

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement describes a proposal to allow an administrator of OpenShift to
have the ability to apply user defined tags to many resources created by OpenShift in AWS.

Note: this enhancement is slightly retroactive. Work has already begun on this. See

>  - https://github.com/openshift/api/pull/864
>  - https://github.com/openshift/cluster-ingress-operator/pull/578

Reviewer 3: ```suggestion
- https://github.com/openshift/api/pull/864
- https://github.com/openshift/cluster-ingress-operator/pull/578
```

## Motivation

Motivations include but are not limited to:

>  - allow admin, compliance, and security teams to keep track of assets and objects created by OpenShift,
>    both at install time and during continuous operation (Day 2)
>  - in a Managed OpenShift environment such as Red Hat OpenShift on AWS (ROSA), allow easy differentiation
>    between Red Hat-managed AWS objects and customer-managed objects
>  - *allow for the restriction of permissions granted to Red Hat in an AWS account by AWS resource tags.
>    For example, see https://issues.redhat.com/browse/SDE-1146 - "IAM users and roles can only operate on resources with specific tags"*

Reviewer 3: ```suggestion
- allow admin, compliance, and security teams to keep track of assets and objects created by OpenShift,
   both at install time and during continuous operation (Day 2)
- in a Managed OpenShift environment such as Red Hat OpenShift on AWS (ROSA), allow easy differentiation
   between Red Hat-managed AWS objects and customer-managed objects
- *allow for the restriction of permissions granted to Red Hat in an AWS account by AWS resource tags.
   For example, see https://issues.redhat.com/browse/SDE-1146 - "IAM users and roles can only operate on resources with specific tags"*
```

### Goals

>  - the administrator or service (in the case of Managed OpenShift) installing OpenShift can pass an arbitrary
>    list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other
>    bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster, and where supported by AWS.
>  - tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
>    after some period of time, apply the tags post-creation.

Reviewer 3: ```suggestion
- the administrator or service (in the case of Managed OpenShift) installing OpenShift can pass an arbitrary
   list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other
   bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster, and where supported by AWS.
- tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
   after some period of time, apply the tags post-creation.
```

### Non-Goals

>  - to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
>    changes the tags stored in the infrastructure resource, behavior is undefined. See below.
>  - to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
>    to prohibit that later.

Reviewer 3: ```suggestion
- to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
   changes the tags stored in the infrastructure resource, behavior is undefined. See below.
- to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
   to prohibit that later.
```

## Proposal

New `experimentalPropagateUserTags` field added to `.platform.aws` of install config to indicate that the user tags should be applied to AWS
resources created by in-cluster operators.

If `experimentalPropagateUserTags` is true, install validation will fail if there is any tag that starts with `kubernetes.io` or `openshift.io`.

> Add a new field `userTags` to `.status.aws` of the `infrastructure.config.openshift.io` type. Tags included in the
> `userTags` field will be applied to new resources created for the cluster. The `userTags` field will be populated by the installer only if the `experimentalPropagateUserTags` field is true.

Reviewer 5: In discussion with the API team, the name of the field in the Infrastructure has been changed to `resourceTags`.
```suggestion
Add a new field `resourceTags` to `.status.aws` of the `infrastructure.config.openshift.io` type. Tags included in the
`resourceTags` field will be applied to new resources created for the cluster. The `resourceTags` field will be populated by the installer only if the `experimentalPropagateUserTags` field is true.
```

---
title: apply-user-defined-tags-to-all-aws-resources-created-by-openshift
authors:
  - "@gregsheremeta"
reviewers:
  - @decarr
  - @bparees
approvers:
  - @decarr 
creation-date: 2021-03-24
last-updated: 2021-04-09
status: implementable
---

# Apply user defined tags to all AWS resources created by OpenShift

## Release Signoff Checklist

> - [ ] Enhancement is `implementable`

Reviewer 6: @gregsheremeta Is this still accurate or have portions of this already been implemented? If so which? can you update this to detail that.

On downgrade:

The status field may remain populated, components may or may not continue to tag newly created resources w/ the additional tags depending on whether or not a given component still has logic to respect the status tags, after the downgrade.

### Version Skew Strategy

TBD

## Implementation History

## Drawbacks

If a customer decides they do not want these tags applied to new resources, there is no clean way to address that need:

1. they would have to get help from support to edit the status field
2. they would have to manually remove the undesired tags from any existing resources

In the future we will want to introduce a spec field where a customer can specify and edit these tags, which will be reflected into status when changed, and we will expect consuming components to reconcile changes to the set of tags by applying net new tags to existing resources (but they will still not remove tags that are dropped from the list of tags).

## Alternatives

> TBD

Reviewer 7: Not sure what other options have already been discussed, but there are several existing auto-tagging projects in the wild.

They generally work by:

- Watching CloudTrail for events
- Adding tags to resources in the events based on some criteria

A simple policy would be something like "tag everything created by X" (and then the installer and all on-cluster components would be sure to use X for AWS interactions)

@estroz and I hacked on something that did this here https://github.com/coreos/grafiti, but I'm sure there are other, newer projects that do something similar.

Author: The main requirement for security-minded customers is that tags be applied to resources atomically at creation-time. Customers set up AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied. Looks like grafiti scans and tags post-creation, which won't work for this use case.

Reviewer 8: It would be good to capture those rejected alternatives, and the reasons, in the body of the doc (either in this draft or an update).

Author: I think it is?
`tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and, after some period of time, apply the tags post-creation.`

Reviewer 8: Sorry, I meant listing the other tool that @ecordell mentioned here in the alternatives section and pointing out that it does not meet the requirements described above.

The idea, at least for me, of having the alternatives section is to capture these sorts of conversations, especially why other approach will _not_ work. That sort of detail is useful to future maintainers.

Reviewer 7: > Customers set up AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied

Do you have any links I could read to learn how this is done?

Reviewer 9: @ecordell checkout https://docs.aws.amazon.com/IAM/latest/UserGuide/access_tags.html and https://docs.aws.amazon.com/IAM/latest/UserGuide/tutorial_attribute-based-access-control.html

Even though it's actually not very common to do it, most security-minded companies would separate AWS accounts and run their workloads using a different IAM user so you can distinguish later on in CloudTrail what resource originated from where.

> AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied. Looks like grafiti scans and tags post-creation, which won't work for this use case.

Not to derail that thread too much, but wouldn't we want introduce such a functionality ourselves first to ensure we always tag these resources? I can see some projects down the road not knowing about the tags and creating resources - would be great to catch that during CI if possible.

---
title: apply-user-defined-tags-to-all-aws-resources-created-by-openshift
authors:
  - "@gregsheremeta"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2021-03-24
last-updated: 2021-03-24
status: implementable
---

# Apply user defined tags to all AWS resources created by OpenShift

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

This enhancement describes a proposal to allow an administrator of OpenShift to

> have the ability to apply user defined tags to all resources created by OpenShift in AWS.

Reviewer 10: apply = overwrite or append to the tags that we supply?

Reviewer 2: it must be append.

Note: this enhancement is slightly retroactive. Work has already begun on this. See
 - https://github.com/openshift/api/pull/864
 - https://github.com/openshift/cluster-ingress-operator/pull/578

## Motivation

Motivations include but are not limited to:

 - allow admin, compliance, and security teams to keep track of assets and objects created by OpenShift,

>    both at install time during continuous operation (Day 2)

Author: fixed

 - in a Managed OpenShift environment such as Red Hat OpenShift on AWS (ROSA), allow easy differentiation
   between Red Hat-managed AWS objects and customer-managed objects
 - *allow for the restriction of permissions granted to Red Hat in an AWS account by AWS resource tags.

>    For example, see SDE-1146 - "IAM users and roles can only operate on resources with specific tags"*

Reviewer 3: Is the SDE Jira public, maybe put a link here? To make "see SDE-1146" easier

Author: done

### Goals

 - the administrator or service (in the case of Managed OpenShift) installing OpenShift can pass an arbitrary
   list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other

>    bootstrapped components will apply those tags to all resources created in AWS, for the lift of the cluster.

Reviewer 5: ```suggestion
   bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster.
```

Author: fixed

 - Installer writes the tags to the infrastructure resource so that Day 2 operations can know what tags to apply

>    to AWS resources for the life of the cluster.

Reviewer 3: Nit, this is more of an implementation detail than a goal, the goal here is to provide a way for Day 2 created AWS resources to also know these tags, better to mention the infrastructure resource in proposal than goals IMO

Author: thanks, removed

 - tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,

>    after some period of time, apply the tags post-creation.

Reviewer 3: How does this work for resources where you can't atomically create tags? Perhaps update this to add the proviso that if a resource supports creating with tags, it isn't acceptable to create and add later.

I'm not sure exactly what resources have this restriction but I know it was brought up on one of the calls

Author: yes, such resources will just be out of scope for now

>    after some period of time, apply the tags post-creation.

Reviewer 11: Have we identified any objects we create for which this is not possible yet?

Author: I have been unable to get a canonical list from AWS

Reviewer 5: Here is a list of supported ec2 resources.

From https://docs.aws.amazon.com/cli/latest/reference/ec2/create-security-group.html,
> The type of resource to tag. Currently, the resource types that support tagging on creation are: capacity-reservation | carrier-gateway | client-vpn-endpoint | customer-gateway | dedicated-host | dhcp-options | egress-only-internet-gateway | elastic-ip | elastic-gpu | export-image-task | export-instance-task | fleet | fpga-image | host-reservation | image | import-image-task | import-snapshot-task | instance | internet-gateway | ipv4pool-ec2 | ipv6pool-ec2 | key-pair | launch-template | local-gateway-route-table-vpc-association | placement-group | prefix-list | natgateway | network-acl | network-interface | reserved-instances |route-table | security-group | snapshot | spot-fleet-request | spot-instances-request | snapshot | subnet | traffic-mirror-filter | traffic-mirror-session | traffic-mirror-target | transit-gateway | transit-gateway-attachment | transit-gateway-multicast-domain | transit-gateway-route-table | volume |vpc | vpc-peering-connection | vpc-endpoint (for interface and gateway endpoints) | vpc-endpoint-service (for AWS PrivateLink) | vpc-flow-log | vpn-connection | vpn-gateway .

Author: done

### Non-Goals

 - to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually

>    changes the tags stores in the infrastructure resource, pre-existing AWS resources are not updated. Only newly-

Reviewer 5: ```suggestion
   changes the tags stored in the infrastructure resource, pre-existing AWS resources are not updated. Only newly-
```

Author: fixed

>    created AWS resources would get the updated tags.

Reviewer 3: I think MAPI was going down the route of reconcile (since we already have the logic for our other tags), would you prefer we didn't so that we conform with others?

Reviewer 10: Do we need to report (not reconcile) - for customers; what tag's are missing or have been altered outside of the OCP platform? 

How will customers, in sensitive situations, know if we have a product bug; or if they have a rouge admin, etc.

Author: I removed this section and clarified the behavior in the risks.

 - to reduce initial scope, we're not implementing this for clouds other than AWS. We shouldn't take any actions to
   prohibit that later, though.

## Proposal

> Add a new field `userTags` to `.spec.aws` of the `infrastructure.config.openshift.io` type. Tags included in the

Reviewer 5: I think the arguments for putting the new field in `.status.platformStatus.aws` instead are compelling. We initially do not want to expose the field as something that the user should expect to either set themselves directly or change post-installation.

Reviewer: This would result in an inability to change/remove this value by the user should the need arise.

Reviewer 5: Correct. That is the intention. To limit the scope of this work, we want to enforce that the user should not change the tags after install, so that we do not have consider the behavior that we want the operators to have in the tag-mutation scenarios. If we add support of the user modifying the tags in the future, then we would add the user-modifiable field to the spec and let a too-be-determined operator (cluster-config-operator?) reconcile from spec to status.

Reviewer 12: > I think the arguments for putting the new field in `.status.platformStatus.aws` instead are compelling.

Agree with this. It really helps with the initial scope of this feature.

Reviewer 11: If in status, is it still true that it should be observable from elsewhere? We've long operated under the guidance that if status were lost (DR perhaps), it should be recovered automatically.

Author: for this experimental pass, status it is

> Add a new field `userTags` to `.spec.aws` of the `infrastructure.config.openshift.io` type. Tags included in the

Reviewer: Is the user allowed to edit anything in this resource today?  Will there be any validation on the supplied values?

Reviewer 12: Validation is something I want to understand too. I mentioned the tag restrictions here in the api PR: https://github.com/openshift/api/pull/864#issuecomment-805656042

Author: for this experimental pass, using status

`userTags` field will be applied to new resources created for the cluster.

Installer will apply these tags to all AWS resources it creates with terraform (e.g. bootstrap and master EC2 instances)

> All operators that create AWS resources (ingress, cloud credential, storage, image registry, machine-api)

Reviewer: Are there any components that exist on the cluster that will create resources in AWS not mentioned here?

I think there will be a strong desire for RH provided bits from OLM to also respect this value, we should possibly add something in enhancement about that.

Reviewer 3: Do we want to mention the service controller for "Service of Type Load Balancer", this is what ingress will be using under the hood but customers will use this too, are we expected to sync the tags to a customer created Service?

Reviewer 13: Kubernetes volume plugin in AWS EBS creates EBS volumes and it will not create EBS volumes with this tag. It's deprecated upstream and it's very hard to pass the tags from OCP API to depths of kube-controller-manager.

We will support it in AWS EBS CSI driver, which is tech preview in 4.8.

will apply these tags to all AWS resources they create.

### User Stories

SDE-1146 - IAM users and roles can only operate on resources with specific tags

> As a security-conscious ROSA customer, I want to restrict the permissions granted to Red Hat in my AWS account by using

Reviewer: What if this user wants different components to utilize different tags, for security purposes?

Reviewer 4: speaking for the image-registry s3 bucket i'd say that's out of scope and if you want that differentiation you bring you own s3 bucket and tell the registry to use it.

Author: good questions -- kicking out of scope for now

AWS resource tags.

### Implementation Details/Notes/Constraints [optional]

Some components, like machine-api and image registry, already have some support to allow adminstrators to
push tags into EC2 instance and S3 buckets, respectively.

> In the case of machine-api, a user has to specify tags on the MachineSet as part of the Day 2 operation.

Reviewer: This is untrue.  Users can modify machinesets during the installation process.

Reviewer 3: It's also important to note here that the installer creates machinesets as part of the install process so _could_ populate the correct fields when it does that

Reviewer 5: In fact, the installer already **does** populate the machines and machinesets that it creates with the additional user-supplied tags.

Reviewer 3: Does that mean there could be conflicts if installer created Machinesets have the same set of tags as going into the infrastructure spec? Or does it populate the tags on the machinesets from a different part of the install config?

Author: removed this section

In the case of image registry, a user has to pre-configure an S3 bucket with the required tags, and use the

> image registry's bring-your-own-bucket feature.

Reviewer 4: this isn't an implementation detail, it's an alternative solution.

Author: thanks, removed

### Risks and Mitigations

To reduce initial scope, tags are applied only at creation time and not reconciled. This behavior is atypical in
Kubernetes and tends to cause confusion for users.

## Design Details

### Open Questions [optional]

 - Since machine-api and image registry already have some support to allow adminstrators to
   push tags into EC2 instance and S3 buckets, respectively, should these be out of scope? Or can we agree
   that it's a nicer user experience to specify the tags one time, at install time, and have everything just
   work after that. Our intention is to change the contract of user tags to state that all resources created
   for the cluster are tagged with the user-provided tags, putting the user-provided tags on equal footing with
   the kubernetes owned tag. If we have to carve out exceptions from there that the user needs to remember to
   fill out themselves, then it dilutes that contract.

 - How are upgrades handled? Is this an install-time only feature that requires a fresh cluster to adopt, or

>    can we somehow enable upgrade users to adopt it?

Reviewer 4: having upgrades able to adopt it falls into the same category as reconciliation for me:  it implies that consuming components are going to watch this config resource in general and update the tags on the aws resource.  Which it was previously stated was out of scope.

If we ever add reconciliation, then upgrade behavior would come for free.

### Test Plan

TBD

### Graduation Criteria

TBD

### Upgrade / Downgrade Strategy

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade in order to keep previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing

>   cluster required to make on upgrade in order to make use of the enhancement?

Reviewer 4: given the lack of reconciliation support, the general behavior (independent of upgrade) is:

if the tags config field is set (and it can only be set after upgrading to a sufficient ocp version since the api won't exist prior to that point in time), any components which create new aws resources after the config field is set would be expected to apply the tags.  Any aws resources that were already created would not get tags added.
